aipmiddleware/validateparent.go

```diff
@@ -0,0 +1,89 @@
+package aipmiddleware
+
+import (
+	"context"
+
+	"go.einride.tech/aip/resourcename"
+	"google.golang.org/genproto/googleapis/api/annotations"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protoreflect"
+)
+
+// ParentValidator validates that response resources are children of requested parent resources.
+type ParentValidator struct {
+	// LogHook is called when parent validation fails, with the failing request/response, parent/name combination.
+	LogHook func(info *grpc.UnaryServerInfo, request, response proto.Message, parent, name string)
+}
+
+// UnaryServerInterceptor implements grpc.UnaryServerInterceptor.
+func (p *ParentValidator) UnaryServerInterceptor(
+	ctx context.Context,
+	request interface{},
+	info *grpc.UnaryServerInfo,
+	handler grpc.UnaryHandler,
+) (interface{}, error) {
+	protoRequest, ok := request.(proto.Message)
+	if !ok {
+		return nil, status.Error(codes.Internal, "parent validator got non-proto request")
+	}
```

Comment:
Not sure when this would happen, or if it is a good idea, but:
If the request is not a proto.Message, should we just pass it on instead of retuning an error?

Reply:
gRPC uses `interface{}` since technically gRPC does not mandate protobuf. My thought here is that this middleware should only be used with proto messages, and the middleware is better off failing loudly than silently not doing anything useful.

Reply:
Fair enough